core/src/main/java/org/apache/iceberg/BaseIncrementalAppendScan.java

```diff
@@ -0,0 +1,179 @@
+    if (context().fromSnapshotId() != null) {
+      if (context().fromSnapshotInclusive()) {
+        // validate the fromSnapshotId is an ancestor of toSnapshotId
+        Preconditions.checkArgument(
+            SnapshotUtil.isAncestorOf(table(), toSnapshotIdInclusive, context().fromSnapshotId()),
+            "Starting snapshot (inclusive) %s is not an ancestor of end snapshot %s",
+            context().fromSnapshotId(), toSnapshotIdInclusive);
+        // for inclusive behavior fromSnapshotIdExclusive is set to the parent snapshot id, which can be null.
+        fromSnapshotIdExclusive = table().snapshot(context().fromSnapshotId()).parentId();
+      } else {
+        // validate the parent snapshot id an ancestor of toSnapshotId
+        Preconditions.checkArgument(
+            SnapshotUtil.isParentAncestorOf(table(), toSnapshotIdInclusive, context().fromSnapshotId()),
+            "Starting snapshot (exclusive) %s is not a parent ancestor of end snapshot %s",
+            context().fromSnapshotId(), toSnapshotIdInclusive);
+      }
+    }
+
+    if (fromSnapshotIdExclusive != null) {
+      Listeners.notifyAll(new IncrementalScanEvent(table().name(), fromSnapshotIdExclusive,
+          toSnapshotIdInclusive, context().rowFilter(), table().schema(), false));
+    } else {
+      Snapshot oldestAncestorSnapshot = SnapshotUtil.oldestAncestor(toSnapshotIdInclusive, table()::snapshot);
+      Listeners.notifyAll(new IncrementalScanEvent(table().name(), oldestAncestorSnapshot.snapshotId(),
+          toSnapshotIdInclusive, context().rowFilter(), table().schema(), true));
+    }
+
+    // appendsBetween handles null fromSnapshotId (exclusive) properly
+    List<Snapshot> snapshots = appendsBetween(table(), fromSnapshotIdExclusive, toSnapshotIdInclusive);
+    if (snapshots.isEmpty()) {
+      return CloseableIterable.empty();
+    }
+
+    return appendFilesFromSnapshots(snapshots);
+  }
+
+  @Override
+  public CloseableIterable<CombinedScanTask> planTasks() {
+    CloseableIterable<FileScanTask> fileScanTasks = planFiles();
+    CloseableIterable<FileScanTask> splitFiles = TableScanUtil.splitFiles(fileScanTasks, targetSplitSize());
+    return TableScanUtil.planTasks(splitFiles, targetSplitSize(), splitLookback(), splitOpenFileCost());
+  }
+
+  private CloseableIterable<FileScanTask> appendFilesFromSnapshots(List<Snapshot> snapshots) {
+    Set<Long> snapshotIds = Sets.newHashSet(Iterables.transform(snapshots, Snapshot::snapshotId));
+    Set<ManifestFile> manifests = FluentIterable
+        .from(snapshots)
+        .transformAndConcat(Snapshot::dataManifests)
+        .filter(manifestFile -> snapshotIds.contains(manifestFile.snapshotId()))
+        .toSet();
+
+    ManifestGroup manifestGroup = new ManifestGroup(tableOps().io(), manifests)
+        .caseSensitive(context().caseSensitive())
+        .select(context().returnColumnStats() ? DataTableScan.SCAN_WITH_STATS_COLUMNS : DataTableScan.SCAN_COLUMNS)
```

Comment:
This block also accesses a lot of information on `context()`. There were some helper methods in `BaseTableScan`. What about moving them to `BaseScan` and using here?

```
  protected Long snapshotId() {
    return context().snapshotId();
  }

  protected boolean colStats() {
    return context().returnColumnStats();
  }

  protected boolean shouldIgnoreResiduals() {
    return context().ignoreResiduals();
  }

  @Override
  public Expression filter() {
    return context().rowFilter();
  }

  @Override
  public boolean isCaseSensitive() {
    return context().caseSensitive();
  }

  protected ExecutorService planExecutor() {
    return context().planExecutor();
  }
```

Reply:
I am not sure adding a bunch help methods like these makes the code simpler

Reply:
These have been already defines in `BaseTableScan` for most common methods as we have like 10+ scans so I was just thinking about moving them to `BaseScan`. Up to you, though. I don't feel strongly.

Reply:
I missed that `BaseTableScan` already has many helper methods like these. I am ok to move the protected helper method to `BaseScan`.  I am not opinionated either way.. A couple of the helper methods (like `filter` and `isCaseSensitive`) from `BaseTableScan` are public and have to stay.

@rdblue do you have any preference?